File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/audit/Auditor.java
package no.nav.pensjon.selvbetjeningopptjening.audit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.event.Level;
import org.springframework.stereotype.Component;

import java.time.ZonedDateTime;

/**
 * Sends info to the auditing system when an internal user (NAV employee) or a fullmektig
 * performs some task on behalf of a regular (external) user.
 * The info is formatted according to ArcSight CEF (Common Event Format).
 */
@Component
public class Auditor {

    private static final Logger audit = LoggerFactory.getLogger("AUDIT_LOGGER");

    public void auditInternalUser(String userId, String onBehalfOfPid) {
        audit.info(cefEntry(userId, "NAV-ansatt", onBehalfOfPid).format());
    }

    public void auditFullmakt(String fullmektigPid, String onBehalfOfPid) {
        audit.info(cefEntry(fullmektigPid, "Fullmektig", onBehalfOfPid).format());
    }

    private static CefEntry cefEntry(String userId, String userType, String onBehalfOfPid) {
        return new CefEntry(
                ZonedDateTime.now().toInstant().toEpochMilli(),
                Level.INFO,
                "audit:read",
                "Datahenting paa vegne av",
                userType + " henter pensjonsopptjening for innbygger",
                userId,
                onBehalfOfPid);
    }
}

File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/tech/security/ingress/impersonal/audit/Auditor.kt
package no.nav.pensjon.selvbetjeningopptjening.tech.security.ingress.impersonal.audit

import mu.KotlinLogging
import no.nav.pensjon.selvbetjeningopptjening.opptjening.Pid
import org.slf4j.event.Level
import org.springframework.stereotype.Component
import java.time.ZonedDateTime

/**
 * Sends info to the auditing system when a user (Nav employee)
 * performs some task on behalf of another person.
 * The info is formatted according to ArcSight CEF (Common Event Format).
 */
@Component
class Auditor(private val ansattIdExtractor: SecurityContextNavIdExtractor) {

    private val log = KotlinLogging.logger("AUDIT_LOGGER")

    fun audit(onBehalfOfPid: Pid, requestUri: String) {
        log.info { cefEntry(ansattIdExtractor.id(), onBehalfOfPid, requestUri).format() }
    }

    private fun cefEntry(ansattId: String, onBehalfOfPid: Pid, requestUri: String) =
        CefEntry(
            timestamp = now(),
            level = Level.INFO,
            deviceEventClassId = DEVICE_EVENT_CLASS_ID,
            name = "Datahenting paa vegne av",
            message = "$USER_TYPE kaller $requestUri",
            sourceUserId = ansattId,
            destinationUserId = onBehalfOfPid.pid
        )

    private companion object {
        private const val DEVICE_EVENT_CLASS_ID = "audit:read"
        private const val USER_TYPE = "Nav-ansatt"

        private fun now() = ZonedDateTime.now().toInstant().toEpochMilli()
    }
}
